File: leetcode/141. Linked List Cycle/answer.c
// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

typedef struct ListNode *Link;

bool hasCycle(struct ListNode *head) {
    if (head == NULL) {
        return false;
    }

    Link slow_p = head, fast_p = head;

    while (slow_p && fast_p && fast_p->next) {
        slow_p = slow_p->next;
        fast_p = fast_p->next->next;
        if (slow_p == fast_p) {
            return 1;
        }
    }
    return 0;
}

File: leetcode/141. Linked List Cycle/answer.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False  # Because there is no ListNode

        visitedNode = set()

        temp = head
        while temp.next is not None:
            visitedNode.add(temp)
            temp = temp.next

            if temp.next in visitedNode:
                return True

        return False
